csrc/iter_visitor.h

```diff
@@ -462,6 +462,14 @@ class DeadCodeRemover : BackwardVisitor {
 
   //! Check whether all uses have been marked dead
   inline bool allUsesDead(Val* val) const {
+    auto fu_it = future_uses_.find(val);
+    if (fu_it != future_uses_.end() && !fu_it->second.empty()) {
+      // Regardless of whether current uses are marked dead, this appears in a
+      // replacement expression, so it has a future live use and we should keep
+      // it.
+      return false;
+    }
```

Comment:
Uses in `future_uses_` may be really dead, so aren't we going to miss real dead uses? In other words, does this change mean this analysis is now less precise and more conservative?

Reply:
Future uses holds Exprs that are used in replacements, so although they are currently dead they will not be after this pass is done. Previously we were not conservative enough and wound up removing Vals which were used only in such future uses.

Reply:
I don't think we will miss removing actual dead code as long as everything registered as a replacement winds up alive. If we register two different replacements for the same Val this might be violated I suppose, but like OptOutMutator, this class assumes we're not doing that.

Reply:
> Future uses holds Exprs that are used in replacements

Is this guaranteed in general?

Reply:
It seems that the code here assumes the particular use case of `DeadCodeRemover`, but I'm assuming the class itself is designed to be more generic as it's defined in iter_visitor.h.